File: week2_decomposition2/3_strongly_connected/strongly_connected.py
#Uses python3
# Remarks:
# Python's built-in List data structure comes bundled with methods to simulate both stack and queue operations.
# Reference at https://docs.python.org/2/tutorial/datastructures.html
# To implement a queue, use collections.deque which was designed to have fast appends and pops from both ends


import sys

sys.setrecursionlimit(200000)

class Directed_Graph:
    def __init__(self):
        self.adjacency_list = None
        self.adjacency_list_reverse = None

    def data_read(self):
        input = sys.stdin.read()
        data = list(map(int, input.split()))
        n, m = data[0:2]
        # [[element] * numcols] * numrows must specify the num cols. otherwise numcols becomes the same
        # When you do [None] * n, Python internally creates a list object of size n and it copies the the same object
        # (here None) (this is the reason, you should use this method only when you are dealing with immutable objects) to all the memory locations.
        # adj = [[] for _ in range(n)], numcols is variable.
        self.adjacency_list = [[] for _ in range(n)]
        self.adjacency_list_reverse = [[] for _ in range(n)]
        data = data[2:]
        edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
        adj = [[] for _ in range(n)]
        for (a, b) in edges:
            self.adjacency_list[a - 1].append(b - 1)
            self.adjacency_list_reverse[b - 1].append(a - 1)
            # The index of the first dimension of the 2D corresponds to the vertex number.
            # Both for original and reverse Adj. List representation.

    def explore(self, v, adj_list, vertex_status, is_cycle, stack_order):
        ''' Check whether there is a cycle containing vertex v.
            Input:
                v(int): Zero-based index for the given vertex.
                adj_list: Adjacency List of the original or reverse version
                vertex_status(vector<string>): "unvisited" if vertex has not been visited yet;
                                "begin" if explore(v) has started;
                                "end" if explore(v) has terminated.
                is_cycle(bool): true if detecting a cycle.
                stack_order(stack<int>): store the vertex numbers in linear order.
                Namely, for edge (u, v), u must show up before v.
            Output:Void.
            Time Complexity: O(# of edges in in the  visited subgraph)
            Space Complexity: O(1). We takes the reference to vector as input.
            Remark: Reference at https://stackoverflow.com/questions/261573/best-algorithm-for-detecting-cycles-in-a-directed-graph
        '''
        vertex_status[v] = "begin"
        for i in range(len(adj_list[v])):
            w = adj_list[v][i]

            if vertex_status[w] == "begin":
                is_cycle[0] = True

            if vertex_status[w] == "unvisited":
                self.explore(w, adj_list,  vertex_status, is_cycle, stack_order)

        vertex_status[v] = "end"
        stack_order.append(v)

    def is_cycle_DFS(self):
        ''' Check whether the graph contains a graph by detecting the back edge.
            Time Complexity: O(|V| + |E|)
            Space Complexity: O(|V|)
        '''
        vertex_status = ["unvisited"] * len(self.adjacency_list)
        # is_cycle = False bool is not mutable
        # The only way around this is to pass a mutable object. Because both references refer to the same object,
        # any changes to the object are reflected in both places.
        # Passing mutable object in python amounts to passing by reference in C++
        # Otherwise, Python will create a new reference.
        is_cycle = [False]
        stack_order = []

        for v in range(len(self.adjacency_list)):
            if vertex_status[v] == "unvisited":
                self.explore(v, self.adjacency_list, vertex_status, is_cycle, stack_order)

        res = 1 if is_cycle[0] else 0
        return res

    def topological_sort(self):
        ''' Linearly order the directed graph.

            A topological sort of a DAG G=(V, E) is a linear ordering of all
            its vertices such that if G contains an edge (u,v), then u appers before v in the ordering.
            Algorithm: Call DFS(G) to compute postorder number.
            As each vertex is finished, insert it to the stack(Last In First Out).
            Time Complexity: O(|V| + |E|)
            Space Complexity: O(|V|)
        '''
        vertex_status = ["unvisited"] * len(self.adjacency_list)
        # is_cycle = False bool is not mutable
        # The only way around this is to pass a mutable object. Because both references refer to the same object,
        # any changes to the object are reflected in both places.
        # Passing mutable object in python amounts to passing by reference in C++
        # Otherwise, Python will create a new reference.
        is_cycle = [False]
        stack_order = []

        for v in range(len(self.adjacency_list)):
            if vertex_status[v] == "unvisited":
                self.explore(v, self.adjacency_list, vertex_status, is_cycle, stack_order)

        return stack_order

    def DFS(self, adj_list):
        '''Depth First Search on the Original Graph.

           Input:
               adj_list: the Adjacency List of the Original or Reverse Graph.
           Return the reversed linear order, ignoring the cycle.
        '''
        vertex_status = ["unvisited"] * len(adj_list)
        # is_cycle = False bool is not mutable
        # The only way around this is to pass a mutable object. Because both references refer to the same object,
        # any changes to the object are reflected in both places.
        # Passing mutable object in python amounts to passing by reference in C++
        # Otherwise, Python will create a new reference.
        is_cycle = [False]
        stack_order = []

        for v in range(len(adj_list)):
            if vertex_status[v] == "unvisited":
                self.explore(v, adj_list, vertex_status, is_cycle, stack_order)

        return stack_order

    def num_sccs(self):
        '''Compute the number of strongly connected components(SCC) in the graph.

           Return: The number SCCs
           Time Complexity:O(|V| + |E|)
           Space Complexity:O(|V|)
        '''
        post_order = []
        post_order = self.DFS(self.adjacency_list_reverse)
        vertex_status = ["unvisited"] * len(self.adjacency_list)
        is_cycle = [False]
        stack_order = []
        number_of_SCCs = 0
        while post_order: # empty list evaluates to false.
            v = post_order.pop()  # return and remove
            if vertex_status[v] == "unvisited":
                self.explore(v, self.adjacency_list, vertex_status, is_cycle, stack_order)
                number_of_SCCs = number_of_SCCs + 1

        return number_of_SCCs

if __name__ == '__main__':
    graph = Directed_Graph()
    graph.data_read()
    print(graph.num_sccs())

    # input = sys.stdin.read()
    # data = list(map(int, input.split()))
    # n, m = data[0:2]
    # data = data[2:]
    # edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    # adj = [[] for _ in range(n)]
    # for (a, b) in edges:
    #     adj[a - 1].append(b - 1)
    # print(number_of_strongly_connected_components(adj))

File: week2_decomposition2/3_strongly_connected/strongly_connected.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
using namespace std;

/* using std::vector; */
/* using std::pair; */
class Directed_Graph
{
  private:
    vector<vector<int>> adjacency_list; // For each vertex, a list of adjacent vertices
    vector<vector<int>> adjacency_list_reverse;

  public:
    void Data_Read()
    {
      size_t n, m;  // n vertices and m edges
      std::cin >> n >> m;
      adjacency_list.resize(n);
      adjacency_list_reverse.resize(n);
      for (size_t i = 0; i < m; i++) {
	int x, y;
	std::cin >> x >> y;  // 1-based
	adjacency_list[x - 1].push_back(y - 1);  // convert to 0-based index
	adjacency_list_reverse[y - 1].push_back(x - 1);
	// The index of the first dimension of the 2D corresponds to the vertex number.
	// Both for original and reverse Adj. List representation.
      }
    }

    Directed_Graph()
    {}
   
    void explore(int v, vector<vector<int>> &adj_list, vector<string> &vertex_status, bool &is_cycle, stack<int> &stack_order)
    {
      /* Check whether there is a cycle containing vertex v. 
	 Input:
	    v(int): Zero-based index for the given vertex.
	    adj_list(vector<vector<int>>): Adjacency List for Original Graph or the Graph with Graph edge reversed.
	    vertex_status(vector<string>): "unvisited" if vertex has not been visited yet;
					    "begin" if explore(v) has started;
					    "end" if explore(v) has terminated.
	    is_cycle(bool): true if detecting a cycle.
	    stack_order(stack<int>): store the linear order.
	 Output:Void.
	 Time Complexity: O(# of edges in in the  visited subgraph)
	 Space Complexity: O(1). We takes the reference to vector as input.
	 Remark: Reference at https://stackoverflow.com/questions/261573/best-algorithm-for-detecting-cycles-in-a-directed-graph
      */

      vertex_status[v] = "begin";
      for(int i(0); i < adj_list[v].size(); i++)
      {
	int w = adj_list[v][i];
	if(vertex_status[w] == "begin")
	{
	  is_cycle = true;
	}
	if(vertex_status[w] == "unvisited")
	{
	  explore(w, adj_list, vertex_status, is_cycle, stack_order);
	}
      }
      vertex_status[v] = "end";
      stack_order.push(v);  // O(1)
    }

    int is_cycle_DFS()
    {
      /* Check whether the graph contains a graph by detecting the back edge.
	 Time Complexity: O(|V| + |E|)
	 Space Complexity: O(|V|)
      */
      vector<string> vertex_status(adjacency_list.size(), "unvisited");
      bool is_cycle = false;
      stack<int> stack_order;
      for(int v(0); v < adjacency_list.size(); v++)
      {
	if(vertex_status[v] == "unvisited")
	{
	  explore(v, adjacency_list, vertex_status, is_cycle, stack_order);
	}
      }

      int res = is_cycle ? 1 : 0;
      return res;
    }

    stack<int> topogical_sort()
    {
      /* Linearly order the directed graph.
	 A topological sort of a DAG G=(V, E) is a linear ordering of all 
	 its vertices such that if G contains an edge (u,v), then u appers before v in the ordering.
	 Algorithm: Call DFS(G) to compute postorder number. 
	 As each vertex is finished, insert it to the stack(Last In First Out).
	 Time Complexity: O(|V| + |E|)
	 Space Complexity:O(|v|)
      */
      vector<string> vertex_status(adjacency_list.size(), "unvisited");
      bool is_cycle = false;
      stack<int> stack_order;
      for(int v(0); v < adjacency_list.size(); v++)
      {
	if(vertex_status[v] == "unvisited")
	{
	  explore(v, adjacency_list, vertex_status, is_cycle, stack_order);
	}
      }
      return stack_order;
    }    

    stack<int> DFS(vector<vector<int>> adj_list)
    {
      /* Depth First Search on the Original Graph.
	 
	 Input: 
	    adj_list: the Adjacency List of the Original or Reverse Graph.
	 Return the reversed linear order, ignoring the cycle.
      */
      vector<string> vertex_status(adj_list.size(), "unvisited");
      bool is_cycle = false;
      stack<int> stack_order;
      for(int v(0); v < adj_list.size(); v++)
      {
	if(vertex_status[v] == "unvisited")
	{
	  explore(v, adj_list, vertex_status, is_cycle, stack_order);
	}
      }

      return stack_order;   // reverse order. smallest postorder first in but last out.
    }

    int num_sccs()
    {
      /* Compute the number of strongly connected components(SCC) in the graph.
	 Return: The number SCCs
	 Time Complexity:O(|V| + |E|)
	 Space Complexity:O(|V|)
      */
      stack<int> post_order;
      post_order = DFS(adjacency_list_reverse); // Largest postorder last in, will first out.

      vector<string> vertex_status(adjacency_list.size(), "unvisited");
      bool is_cycle = false;
      stack<int> stack_order;
      int number_of_SCCs(0);
      while(!post_order.empty())
      {
	int v = post_order.top(); // Largest post_order comes out first.
	post_order.pop();
	if(vertex_status[v] == "unvisited")
	{
	  explore(v, adjacency_list, vertex_status, is_cycle, stack_order);
	  number_of_SCCs ++;
	}
      }

      return number_of_SCCs; 
    }
};



int main() {
  // https://stackoverflow.com/questions/9490349/instantiate-class-with-or-without-parentheses
  Directed_Graph graph;
  graph.Data_Read();
  cout << graph.num_sccs() << endl;
  /* std::cout << number_of_strongly_connected_components(adj); */
}
